File: src/main/java/com/tobe/healthy/lessonhistory/repository/LessonHistoryCommentRepositoryImpl.java
package com.tobe.healthy.lessonhistory.repository;

import static com.tobe.healthy.lessonhistory.domain.QLessonHistoryComment.*;

import org.springframework.stereotype.Repository;
import org.springframework.util.ObjectUtils;

import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQueryFactory;
import com.tobe.healthy.lessonhistory.domain.LessonHistoryComment;

import lombok.RequiredArgsConstructor;

@Repository
@RequiredArgsConstructor
public class LessonHistoryCommentRepositoryImpl implements LessonHistoryCommentRepositoryCustom {

	private final JPAQueryFactory queryFactory;

	@Override
	public int findTopComment(Long lessonHistoryId, Long lessonHistoryCommentId) {
		Integer result = queryFactory
			.select(lessonHistoryComment.order.max().add(1))
			.from(lessonHistoryComment)
			.where(
				lessonHistoryIdEq(lessonHistoryId),
				parentCommentIdEq(lessonHistoryCommentId)
			)
			.fetchOne();
		return result != null ? result : 1;
	}

	@Override
	public LessonHistoryComment findLessonHistoryCommentWithFiles(Long lessonHistoryCommentId, Long writerId) {
		return queryFactory
			.select(lessonHistoryComment)
			.from(lessonHistoryComment)
			.leftJoin(lessonHistoryComment.files).fetchJoin()
			.where(
				lessonHistoryComment.id.eq(lessonHistoryCommentId),
				lessonHistoryComment.delYn.eq(false),
				lessonHistoryComment.writer.id.eq(writerId)
			)
			.fetchOne();
	}

	@Override
	public LessonHistoryComment findCommentById(Long lessonHistoryCommentId) {
		return queryFactory
			.select(lessonHistoryComment)
			.from(lessonHistoryComment)
			.where(
				lessonHistoryComment.id.eq(lessonHistoryCommentId),
				lessonHistoryComment.delYn.eq(false)
			)
			.fetchOne();
	}

	@Override
	public LessonHistoryComment findById(Long lessonHistoryCommentId, Long writerId) {
		return queryFactory
			.select(lessonHistoryComment)
			.from(lessonHistoryComment)
			.where(
				lessonHistoryComment.id.eq(lessonHistoryCommentId),
				lessonHistoryComment.delYn.eq(false),
				lessonHistoryComment.writer.id.eq(writerId)
			)
			.fetchOne();
	}

	private BooleanExpression parentCommentIdEq(Long lessonHistoryCommentParentId) {
		if (ObjectUtils.isEmpty(lessonHistoryCommentParentId)) {
			return null;
		}
		return lessonHistoryComment.parent.id.eq(lessonHistoryCommentParentId);
	}

	private BooleanExpression parentCommentIdIsNull() {
		return lessonHistoryComment.parent.isNull();
	}

	private BooleanExpression lessonHistoryIdEq(Long lessonHistoryId) {
		return lessonHistoryComment.lessonHistory.id.eq(lessonHistoryId);
	}
}

File: src/main/kotlin/com/tobe/healthy/lessonHistory/repository/LessonHistoryCommentRepositoryImpl.kt
package com.tobe.healthy.lessonHistory.repository

import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import com.tobe.healthy.lessonHistory.domain.entity.QLessonHistoryComment.lessonHistoryComment
import org.springframework.stereotype.Repository

@Repository
class LessonHistoryCommentRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : LessonHistoryCommentRepositoryCustom {

    override fun findTopComment(lessonHistoryId: Long): Int {
        val result = queryFactory
            .select(lessonHistoryComment.order.max().add(1))
            .from(lessonHistoryComment)
            .where(lessonHistoryIdEq(lessonHistoryId), parentCommentIdIsNull())
            .fetchOne() ?: 1
        return result;
    }

    override fun findTopComment(lessonHistoryId: Long, lessonHistoryCommentParentId: Long): Int {
        val result = queryFactory
            .select(lessonHistoryComment.order.max().add(1))
            .from(lessonHistoryComment)
            .where(lessonHistoryIdEq(lessonHistoryId), parentCommentIdEq(lessonHistoryCommentParentId))
            .fetchOne() ?: 1
        return result;
    }

    private fun parentCommentIdEq(lessonHistoryCommentParentId: Long): BooleanExpression? =
        lessonHistoryComment.parentId.id.eq(lessonHistoryCommentParentId)

    private fun parentCommentIdIsNull(): BooleanExpression =
        lessonHistoryComment.parentId.isNull

    private fun lessonHistoryIdEq(lessonHistoryId: Long): BooleanExpression =
        lessonHistoryComment.lessonHistory.id.eq(lessonHistoryId)
}
